File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.py
MAX_INT = 2 ** 31 - 1
MIN_INT = -1 * 2 ** 31 

VALID_CHARS = {
    "0": 0,
    "1": 1, 
    "2": 2, 
    "3": 3, 
    "4": 4, 
    "5": 5, 
    "6": 6, 
    "7": 7, 
    "8": 8,
    "9": 9,
}

class Solution:
    def myAtoi(self, s: str) -> int:
        while len(s) > 0 and s[0] == " ":
            s = s[1:]
                
        multiplier = 1        
        if len(s) > 0:
            if s[0] == "-":
                multiplier = -1
                s = s[1:]
            elif s[0] == "+":
                s = s[1:]
        
        res = 0
        while len(s) > 0 and s[0] in VALID_CHARS:
            res = (res * 10) + VALID_CHARS[s[0]]
            s = s[1:]
            
        res *= multiplier    
        
        if res < MIN_INT:
            return MIN_INT
            
        if MAX_INT < res:
            return MAX_INT
            
        return res 

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.go
import (
    "math"
)

var runeToInt = map[rune]int{
    '0': 0,
    '1': 1,
    '2': 2,
    '3': 3,
    '4': 4,
    '5': 5,
    '6': 6,
    '7': 7,
    '8': 8,
    '9': 9,
}

func myAtoi(s string) int {
    for len(s) > 0 && rune(s[0]) == ' ' {
        s = s[1:]
    }
    
    if len(s) == 0 {
        return 0
    }
    
    multiplier := 1
    if s[0] == '-' {
        multiplier = -1
        s = s[1:]
    } else if s[0] == '+' {
        s = s[1:]
    }
    
    result := 0
    for len(s) > 0 {
        v, ok := runeToInt[rune(s[0])]
        if !ok {
            break
        }
        
        limit, ok := willOverflow(result * multiplier, v)
        if ok {
            return limit
        }
        
        result = result * 10 + v
        s = s[1:]
    }
    
    return result * multiplier
}

// willOverflow returns the limit exceeded,
// and ok if overflow occurs
func willOverflow(res, add int) (int, bool) {
    if res > math.MaxInt32 / 10 + 7 - add {
        return math.MaxInt32, true 
    } 
    if res < math.MinInt32 / 10 - 8 + add{
        return math.MinInt32, true 
    }
    
    return 0, false
}